### ISSUE TITLE
Is the reward computed via dof(axis angle). I add the code in ASE code, but  the reword becomes negative. I wonder is it  right?

### ISSUE THREAD

**Issue by interestingzhuo** [ZhengyiLuo/UniversalHumanoidControl#23]
Is the reward computed via dof(axis angle). I add the code in ASE code, but  the reword becomes negative. I wonder is it  right?
```
dof_force_tensor = self.gym.acquire_dof_force_tensor(self.sim)
self.gym.refresh_dof_force_tensor(self.sim)
self._dof_force = gymtorch.wrap_tensor(dof_force_tensor).view(self.num_envs, dofs_per_env)    
def get_energy(self):
        dof_force, dof_vel, dof_obs_size, dof_offsets = self._dof_force, self._dof_vel, self._dof_obs_size, self._dof_offsets
        num_joints = len(dof_offsets) - 1
        num_env = dof_force.shape[0]
        energy = torch.zeros(num_env, device = dof_force.device)

        for j in range(num_joints):
            dof_offset = dof_offsets[j]
            dof_size = dof_offsets[j + 1] - dof_offsets[j]
            force = dof_force[:, dof_offset:(dof_offset + dof_size)]
            vel = dof_vel[:, dof_offset:(dof_offset + dof_size)]
            energy += torch.mul(force, vel).sum(dim = -1)**2
        energy =  0.00005*energy
        return energy

rew_buf += -1 * self.energy_buf
```

_Originally posted by @interestingzhuo in https://github.com/ZhengyiLuo/UniversalHumanoidControl/issues/21#issuecomment-1610825890_

**Comment by ZhengyiLuo** [comment]
It should be negative. It is more a energy penalty than an reward.

**Comment by interestingzhuo** [comment]
> 

Is the reward computed via dof(axis angle)? I add the code in ASE code, but  the reword becomes negative(about -2100). and the whole reward will be negative. I wonder is it  right?

**Comment by ZhengyiLuo** [comment]
Hope this problem has been resolved, but here is the code I used in PHC:

`power = torch.abs(torch.multiply(self.dof_force_tensor, self._dof_vel)).sum(dim=-1) `

**Comment by interestingzhuo** [comment]
> Hope this problem has been resolved, but here is the code I used in PHC:
> 
> `power = torch.abs(torch.multiply(self.dof_force_tensor, self._dof_vel)).sum(dim=-1) `

self.dof_force_tensor -> self._dof_force?

**Comment by ZhengyiLuo** [comment]
They probably refer to the same thing based on your code base 😁
